### PR TITLE
drivers: gpio: ifx: Update GPIO driver to use PDL instead of HAL

### PR DESCRIPTION
**Pull request by sreeramIfx**
Update the GPIO driver to use PDL API instead of HAL API to fix the issue with configuring interrupts on both edges

### CODE REVIEW

File: drivers/gpio/gpio_ifx_cat1.c [hunk 1]
```diff
@@ -202,27 +165,28 @@ static uint32_t gpio_cat1_get_pending_int(const struct device *dev)
...
-static void gpio_event_callback(void *callback_arg, cyhal_gpio_event_t event)
+static void gpio_isr_handler(const struct device *dev)
 {
-	ARG_UNUSED(event);
-	uint32_t port_num = CYHAL_GET_PORT((uint32_t) callback_arg);
-	uint32_t pin_num = CYHAL_GET_PIN((uint32_t) callback_arg);
-	const struct device *dev = port_dev_obj[port_num];
+	const struct gpio_cat1_config *const cfg = dev->config;
+	GPIO_PRT_Type *const base = cfg->regs;
+	uint32_t pins = GPIO_PRT_INTR(base);
```
Comment (npal-cy) [inline drivers/gpio/gpio_ifx_cat1.c:172]: you should use GPIO_PRT_INTR_MASKED instead GPIO_PRT_INTR, to check masked interrupts.

File: drivers/gpio/gpio_ifx_cat1.c [hunk 2]
```diff
@@ -202,27 +165,28 @@ static uint32_t gpio_cat1_get_pending_int(const struct device *dev)
...
+
+	for (uint8_t i = 0; i < 8; i++) {
```
Comment (npal-cy) [inline drivers/gpio/gpio_ifx_cat1.c:174]: 8 ->  CY_GPIO_PINS_MAX

**Review by npal-cy — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_ifx_cat1.c (hunk 1 above)
Comment (sreeramIfx) [reply drivers/gpio/gpio_ifx_cat1.c]: updated..

File: drivers/gpio/gpio_ifx_cat1.c (hunk 2 above)
Comment (sreeramIfx) [reply drivers/gpio/gpio_ifx_cat1.c]: updated..

File: drivers/gpio/gpio_ifx_cat1.c [hunk 3]
```diff
@@ -50,93 +48,58 @@ struct gpio_cat1_data {
...
-	 * IOSS_GPIO_GPIO_PORT_NR for the respective categories, but using
-	 * the macro in LISTIFY causes build failures
-	 */
-	#if CONFIG_SOC_FAMILY_INFINEON_CAT1A
-		LISTIFY(15, GET_DEV_OBJ_FROM_LIST, (,))
-	#elif CONFIG_SOC_FAMILY_INFINEON_CAT1B
-		LISTIFY(6, GET_DEV_OBJ_FROM_LIST, (,))
-	#endif
-};
+#define GET_PORT_NUM(dev)                                                                          \
```
Comment (ifyall) [inline drivers/gpio/gpio_ifx_cat1.c:51]: I would be interested in a comment that explains why this math works.

**Review by ifyall — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_ifx_cat1.c [hunk 4]
```diff
@@ -50,93 +48,58 @@ struct gpio_cat1_data {
...
+			drive_mode = CY_GPIO_DM_STRONG;
 			pin_val = (flags & GPIO_OUTPUT_INIT_HIGH) ? true : false;
 		}
 		break;
 
 	case GPIO_DISCONNECTED:
-		/* Handle this after calling cyhal_gpio_init(), otherwise it will cause an assert
-		 * from HAL for freeing an uninitialized pin
-		 */
+		Cy_GPIO_SetInterruptMask(Cy_GPIO_PortToAddr(GET_PORT_NUM(dev)), pin, 0);
```
Comment (npal-cy) [inline drivers/gpio/gpio_ifx_cat1.c:90]: let's remove usage of GET_PORT_NUM(dev)

```
...
const struct gpio_cat1_config *const cfg = dev->config;
GPIO_PRT_Type *const base = cfg->regs;
	
...
Cy_GPIO_SetInterruptMask(base , pin, 0);

```

File: drivers/gpio/gpio_ifx_cat1.c [hunk 5]
```diff
@@ -50,93 +48,58 @@ struct gpio_cat1_data {
...
-		cyhal_gpio_free(gpio_pin);
-		status = cyhal_gpio_init(gpio_pin, gpio_dir, gpio_mode, pin_val);
-	}
-
-	if (flags & GPIO_DISCONNECTED) {
-		cyhal_gpio_free(gpio_pin);
-	}
-
-	return (status == CY_RSLT_SUCCESS) ? 0 : -EIO;
+	Cy_GPIO_Pin_FastInit(Cy_GPIO_PortToAddr(GET_PORT_NUM(dev)), pin, drive_mode,
```
Comment (npal-cy) [inline drivers/gpio/gpio_ifx_cat1.c:99]: uses `base` instead Cy_GPIO_PortToAddr(GET_PORT_NUM(dev))

File: drivers/gpio/gpio_ifx_cat1.c [hunk 6]
```diff
@@ -202,27 +165,28 @@ static uint32_t gpio_cat1_get_pending_int(const struct device *dev)
...
-	ARG_UNUSED(event);
-	uint32_t port_num = CYHAL_GET_PORT((uint32_t) callback_arg);
-	uint32_t pin_num = CYHAL_GET_PIN((uint32_t) callback_arg);
-	const struct device *dev = port_dev_obj[port_num];
+	const struct gpio_cat1_config *const cfg = dev->config;
+	GPIO_PRT_Type *const base = cfg->regs;
+	uint32_t pins = GPIO_PRT_INTR_MASKED(base);
+
+	for (uint8_t i = 0; i < CY_GPIO_PINS_MAX; i++) {
+		Cy_GPIO_ClearInterrupt(Cy_GPIO_PortToAddr(GET_PORT_NUM(dev)), i);
```
Comment (npal-cy) [inline drivers/gpio/gpio_ifx_cat1.c:175]: let's use `base` instead Cy_GPIO_PortToAddr(GET_PORT_NUM(dev))

File: drivers/gpio/gpio_ifx_cat1.c [hunk 7]
```diff
@@ -231,51 +195,26 @@ static int gpio_cat1_pin_interrupt_configure(const struct device *dev, gpio_pin_
...
-
-	/* Find index of free callback data structure */
-	uint32_t index;
-
-	for (index = 0u; index < cfg->ngpios; index++) {
-		if ((cb_data_ptr[index].callback == NULL) || (cb_data_ptr[index].pin == gpio_pin)) {
-			break;
-		}
-	}
+	Cy_GPIO_SetInterruptEdge(Cy_GPIO_PortToAddr(GET_PORT_NUM(dev)), pin, trig_pdl);
```
Comment (npal-cy) [inline drivers/gpio/gpio_ifx_cat1.c:213]: let's use `base` instead Cy_GPIO_PortToAddr(GET_PORT_NUM(dev))

File: drivers/gpio/gpio_ifx_cat1.c (hunk 3 above)
Comment (npal-cy) [reply drivers/gpio/gpio_ifx_cat1.c:51]: let's remove GET_PORT_NUM, we should use  `base` instead Cy_GPIO_PortToAddr(GET_PORT_NUM(dev)). GET_PORT_NUM was needed for cyhal.

**Review by npal-cy — CHANGES_REQUESTED** [review]

**Review by ifyall — APPROVED** [review]

**Review by npal-cy — APPROVED** [review]

**Pull request merged**